<h1>Chapter 2 - Tokens and Token Embeddings</h1>

This notebook is for Chapter 2 of the [Hands-On Large Language Models](https://www.amazon.com/Hands-Large-Language-Models-Understanding/dp/1098150961) book by [Jay Alammar](https://www.linkedin.com/in/jalammar) and [Maarten Grootendorst](https://www.linkedin.com/in/mgrootendorst/).

You should use a GPU to run the examples in this notebook. In Google Colab, go to

**Runtime > Change runtime type > Hardware accelerator > GPU > T4 GPU.**

Uncomment and run the following codeblock to install the dependencies.

In [ ]:
%%capture
!pip install transformers>=4.41.2 sentence-transformers>=3.0.1 gensim>=4.3.2 scikit-learn>=1.5.0 accelerate>=0.31.0

# Downloading and Running An LLM

Load the Phi-3-mini-4k-instruct model onto the GPU, same as Chapter 1.

In [ ]:
from transformers import AutoModelForCausalLM, AutoTokenizer

# Load model and tokenizer
model = AutoModelForCausalLM.from_pretrained(
    "microsoft/Phi-3-mini-4k-instruct",
    device_map="cuda",
    torch_dtype="auto",
    trust_remote_code=True,
)
tokenizer = AutoTokenizer.from_pretrained("microsoft/Phi-3-mini-4k-instruct")

config.json:   0%|          | 0.00/967 [00:00<?, ?B/s]

configuration_phi3.py:   0%|          | 0.00/11.2k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/microsoft/Phi-3-mini-4k-instruct:
- configuration_phi3.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


modeling_phi3.py:   0%|          | 0.00/73.2k [00:00<?, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/microsoft/Phi-3-mini-4k-instruct:
- modeling_phi3.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


model.safetensors.index.json:   0%|          | 0.00/16.5k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/4.97G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/2.67G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/181 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/3.44k [00:00<?, ?B/s]

tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.94M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/306 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/599 [00:00<?, ?B/s]

Unlike chapter 1, this time _don't_ load tranformers.pipline to encapsulate the model, tokenizer, and text-generation process into a single pipeline function. This way, we can explore them separately.

In the code cell below, the prompt is converted to tokens, the model processes the tokens and returns its generative result as tokens, and those tokens are converted back to readable text: **Subject: Sincere Apologies for the Gardening Mishap... Dear**.

In [ ]:
prompt = "Write an email apologizing to Sarah for the tragic gardening mishap. Explain how it happened.<|assistant|>"

# Tokenize the input prompt and return the tokens as PyTorch (pt) tensors.
# `.input_ids.to()` converts the tokens to a CUDA tensor so they can be
# processed on the GPU.
input_ids = tokenizer(prompt, return_tensors="pt").input_ids.to("cuda")

# Generate text from the prompt tokens, stopping at 20 tokens of response.
generation_output = model.generate(
  input_ids=input_ids,
  max_new_tokens=20
)

# Decode the response tokens back into readable text.
print(tokenizer.decode(generation_output[0]))

The attention mask is not set and cannot be inferred from input because pad token is same as eos token. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
The `seen_tokens` attribute is deprecated and will be removed in v4.41. Use the `cache_position` model input instead.


Write an email apologizing to Sarah for the tragic gardening mishap. Explain how it happened.<|assistant|> Subject: Sincere Apologies for the Gardening Mishap


Dear


`input_ids` is a tensor of index values for reference table inside the tokenizer.

In [ ]:
print(input_ids)

tensor([[14350,   385,  4876, 27746,  5281,   304, 19235,   363,   278, 25305,
           293, 16423,   292,   286,   728,   481, 29889, 12027,  7420,   920,
           372,  9559, 29889, 32001]], device='cuda:0')


Here's how it tokenized the prompt. Use the `tokenizer.decode()` method on the prompt tokens. Notice the first token is "\<s\>". That's a special indicator token identifying the beginning of text. Tokens are not always words or even stems. Punctuation marks are their own tokens.

In [ ]:
for id in input_ids[0]:
   print(tokenizer.decode(id))

Write
an
email
apolog
izing
to
Sarah
for
the
trag
ic
garden
ing
m
ish
ap
.
Exp
lain
how
it
happened
.
<|assistant|>


Take a look at `generation_output`. The tokens up through "32001" are the input tokens. Then "3323" - "799" are the output tokens.

In [ ]:
generation_output

tensor([[14350,   385,  4876, 27746,  5281,   304, 19235,   363,   278, 25305,
           293, 16423,   292,   286,   728,   481, 29889, 12027,  7420,   920,
           372,  9559, 29889, 32001,  3323,   622, 29901,   317,  3742,   406,
          6225, 11763,   363,   278, 19906,   292,   341,   728,   481,    13,
            13,    13, 29928,   799]], device='cuda:0')

`tokenizer.decode()` translates individual tokens or a list of them. Here are the first three again.

In [ ]:
print(tokenizer.decode(3323))
print(tokenizer.decode(622))
print(tokenizer.decode([3323, 622]))
print(tokenizer.decode(29901))

Sub
ject
Subject
:


So how does a tokenizer tokenize text? There are several tokenization methods. Byte pair encoding (BPE, used in many GPT models), and WordPiece (used by BERT) are two. After selecting vocabulary size and other settings, the tokenizer is trainined on a corpus of text to establish the best vocabulary.

Tokenization could be done at the word level. Word2vec does word tokenization. But this results in many tokens with minimal differences (e.g., apology, apologize, apologetic, apologist), and the increased potential for encountering new words. At the other extreme, you could tokenize by character, but that turns out to be difficult to model. There is another method to be aware of called _byte tokens_ that represent unicode characters.

# Comparing Trained LLM Tokenizers

Tokenizers differ in their tokenization method, parameters and specialization tokens used to initialize, and their training dataset. Below we define a function that tokenizes a `sentence` using the requested `tokenizer_name`. The `colors_list` helps visualize the different tokens.

In [ ]:
# from transformers import AutoModelForCausalLM, AutoTokenizer

colors_list = [
    '102;194;165', '252;141;98', '141;160;203',
    '231;138;195', '166;216;84', '255;217;47'
]

def show_tokens(sentence, tokenizer_name):
    tokenizer = AutoTokenizer.from_pretrained(tokenizer_name)
    token_ids = tokenizer(sentence).input_ids
    for idx, t in enumerate(token_ids):
        print(
            f'\x1b[0;30;48;2;{colors_list[idx % len(colors_list)]}m' +
            tokenizer.decode(t) +
            '\x1b[0m',
            end=' '
        )

We'll use the following text to show how tokenizers treat capitalization, non-English language, emojis, programming code, and numbers.

In [ ]:
text = """
English and CAPITALIZATION
🎵 鸟
show_tokens False None elif == >= else: two tabs:"    " Three tabs: "       "
12.0*50=600
"""

First up is **BERT base model (uncased)** ([Hugging Face](https://huggingface.co/google-bert/bert-base-uncased), 2018). It uses WordPiece to tokenize. Its vocabular is 30,522. Unknown tokens show up as \[UNK\]. The emoji and Chinese chars are \[UNK\]. Newline breaks are gone - the model is blind to information implied by newlines.

In [ ]:
show_tokens(text, "bert-base-uncased")

[CLS] english and capital ##ization [UNK] [UNK] show _ token ##s false none eli ##f = = > = else : two tab ##s : " " three tab ##s : " " 12 . 0 * 50 = 600 [SEP] 

Next is **BERT base model (cased) ([Hugging Face](https://huggingface.co/google-bert/bert-base-cased), 2018). It also uses WordPiece to tokenize. Its vocabulary is slightly smaller, 28,996. This encoder includes uppercase tokens.

In [ ]:
show_tokens(text, "bert-base-cased")

tokenizer_config.json:   0%|          | 0.00/49.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/436k [00:00<?, ?B/s]

[CLS] English and CA ##PI ##TA ##L ##I ##Z ##AT ##ION [UNK] [UNK] show _ token ##s F ##als ##e None el ##if = = > = else : two ta ##bs : " " Three ta ##bs : " " 12 . 0 * 50 = 600 [SEP] 

**GPT-2** ([Hugging Face](https://huggingface.co/openai-community/gpt2), 2019) uses bye-pair encoding (BPE). The vocabulary size is twice as large, 50,257. GPT-2 recognizes newlines. It recognizes the emoji and Chinese character, even though the decoded tokens don't display it. This model also recognizes tabs and whitespace - important for code like python.

In [ ]:
show_tokens(text, "gpt2")

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]


 English  and  CAP ITAL IZ ATION 
 � � �  � � � 
 show _ t ok ens  False  None  el if  ==  >=  else :  two  tabs :"        "  Three  tabs :  "              " 
 12 . 0 * 50 = 600 
 

**Flan-T5** ([Hugging Face](https://huggingface.co/google/flan-t5-xxl), 2022) tokenizes with SentencePiece. Vocabulary size is 32,100. Notice this one doesn't recognize newlines or whitespace - it wouldn't work great for computer code.

In [ ]:
show_tokens(text, "google/flan-t5-small")

tokenizer_config.json:   0%|          | 0.00/2.54k [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.42M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/2.20k [00:00<?, ?B/s]

English and CA PI TAL IZ ATION  <unk>  <unk> show _ to ken s Fal s e None  e l if = = > = else : two tab s : " " Three tab s : " " 12. 0 * 50 = 600  </s> 

**GPT-4** (2023) has a large (100,000+) vocabulary. It introduces "fill in the middle" tokens that help LLMs use text before an after.

In [ ]:
# The official is `tiktoken` but this the same tokenizer on the HF platform
show_tokens(text, "Xenova/gpt-4")

tokenizer_config.json:   0%|          | 0.00/460 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/2.01M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/917k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/4.23M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/98.0 [00:00<?, ?B/s]


 English  and  CAPITAL IZATION 
 � � �  � � � 
 show _tokens  False  None  elif  ==  >=  else :  two  tabs :"      "  Three  tabs :  "         "
 12 . 0 * 50 = 600 
 

**StarCoder2** ([HuggingFace](https://huggingface.co/bigcode/starcoder2-15b), 2024) has a vocabular size of 48,152. It is focused o generating code.

In [ ]:
# You need to request access before being able to use this tokenizer
show_tokens(text, "bigcode/starcoder2-15b")

config.json:   0%|          | 0.00/803 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/7.88k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/777k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/442k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.06M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/958 [00:00<?, ?B/s]


 English  and  CAPITAL IZATION 
 � � �   � � 
 show _ tokens  False  None  elif  ==  >=  else :  two  tabs :"      "  Three  tabs :  "         " 
 1 2 . 0 * 5 0 = 6 0 0 
 

**Galactica** ([Hugging Face](https://huggingface.co/facebook/galactica-1.3b)) has a vocabulary size of 50,000. It is focused on science. Trained on journal articles and other reference material, it includes special tokens for citations, DNA, and math.

In [ ]:
show_tokens(text, "facebook/galactica-1.3b")

tokenizer_config.json:   0%|          | 0.00/166 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.14M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/3.00 [00:00<?, ?B/s]


 English  and  CAP ITAL IZATION 
 � � � �  � � � 
 show _ tokens  False  None  elif   ==   > =  else :  two  t abs : "      "  Three  t abs :   "         " 
 1 2 . 0 * 5 0 = 6 0 0 
 

Last up is **Phi-3 (and Llama 2)** ([Hugging Face](https://huggingface.co/microsoft/Phi-3-mini-4k-instruct). Recall that Phi-3 uses the tokenizer from Llama 2. It uses the BPE method. It's vocabulary size is 32,000. It has special tokens to recognize turns in a conversation (user, assistant, and system).

In [ ]:
show_tokens(text, "microsoft/Phi-3-mini-4k-instruct")

 
 English and C AP IT AL IZ ATION 
 � � � �  � � � 
 show _ to kens False None elif == >= else : two tabs :"    " Three tabs : "       " 
 1 2 . 0 * 5 0 = 6 0 0 
 

# Contextualized Word Embeddings From a Language Model (Like BERT)

After the tokenizer is initialized and trained, it is used to train the LLM. This is why LLMs and tokenizers are always linked. The LLM has an embedding vector for each token in the tokenizer vocabulary. Initially the embedding vectors are initialized to something random, then the training process optimizes the values.

Instead of representing each token with a static vector, LLMs created contextualized word embeddings. Let's see how this is done using the DeBERTa v3 LLM. `tokenizer()` tokenizes "Hello world", and `model()` embeds each token into a 384-value vector.

In [1]:
from transformers import AutoModel, AutoTokenizer

# Load a tokenizer
tokenizer = AutoTokenizer.from_pretrained("microsoft/deberta-base")

# Load a language model
model = AutoModel.from_pretrained("microsoft/deberta-v3-xsmall")

# Tokenize the sentence
tokens = tokenizer('Hello world', return_tensors='pt')

# Process the tokens
output = model(**tokens)[0]

tokenizer_config.json:   0%|          | 0.00/52.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/474 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


config.json:   0%|          | 0.00/578 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/241M [00:00<?, ?B/s]

`tokens` is a multidimensional array. The first dimension is the batch dimension used in cases (like training). The other dimensions are 4 tokens, each embedded in a 384-value vector.

In [ ]:
output.shape

torch.Size([1, 4, 384])

Here are the four tokens, translated back into English.

In [2]:
for token in tokens['input_ids'][0]:
    print(tokenizer.decode(token))

[CLS]
Hello
 world
[SEP]


And here are the four vectors.

In [3]:
output

tensor([[[-3.4816,  0.0861, -0.1819,  ..., -0.0612, -0.3911,  0.3017],
         [ 0.1898,  0.3208, -0.2315,  ...,  0.3714,  0.2478,  0.8048],
         [ 0.2071,  0.5036, -0.0485,  ...,  1.2175, -0.2292,  0.8582],
         [-3.4278,  0.0645, -0.1427,  ...,  0.0658, -0.4367,  0.3834]]],
       grad_fn=<NativeLayerNormBackward0>)

# Text Embeddings (For Sentences and Whole Documents)

LLMs usually operate on entire sentences, paragraphs, or whole documents. Special models produce _text embeddings_, a single vector the represents pieces of text. One way to produce a text embedding is to average over the token embeddings. However, high quality text embedding models have been developed. Use the sentence-transformers package to load publicly available text embedding models. Let's try this with all-mpnet-base-v2 ([Hugging Face](https://huggingface.co/sentence-transformers/all-mpnet-base-v2)). It maps sentences and paragraphs to a 768 dimensional dense vector space.

In [4]:
from sentence_transformers import SentenceTransformer

# Load model
model = SentenceTransformer('sentence-transformers/all-mpnet-base-v2')

# Convert text to text embeddings
vector = model.encode("Best movie ever!")

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.6k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/571 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/438M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/363 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

 The sentence is now enoded into a 768-value vector.

In [ ]:
vector.shape

(768,)

# Word Embeddings Beyond LLMs


In [5]:
import gensim.downloader as api

# Download embeddings (66MB, glove, trained on wikipedia, vector size: 50)
# Other options include "word2vec-google-news-300"
# More options at https://github.com/RaRe-Technologies/gensim-data
model = api.load("glove-wiki-gigaword-50")

[==================================================] 100.0% 66.0/66.0MB downloaded


In [ ]:
model.most_similar([model['king']], topn=11)

[('king', 1.0000001192092896),
 ('prince', 0.8236179351806641),
 ('queen', 0.7839043140411377),
 ('ii', 0.7746230363845825),
 ('emperor', 0.7736247777938843),
 ('son', 0.766719400882721),
 ('uncle', 0.7627150416374207),
 ('kingdom', 0.7542161345481873),
 ('throne', 0.7539914846420288),
 ('brother', 0.7492411136627197),
 ('ruler', 0.7434253692626953)]

Like LLMs, word2vec is trained on text examples. The algorithm uses a sliding window to generate training examples. The embeddings are generated from a classification task using a neural network to predict whether words commonly appear in the same context. For example, given the phrase "Thou shalt not make a machine in the likeness of a human mind", the algorith might form a sliding window 2 to the left and right of each word, prediting "1" for "Thou", "shalt", "Make", and "a" for the middle word "not". To create negative examples, the text would include negative examples too - random words labeled "0". After creating millions (or even billions) of examples, an randomly initialized embedding vector is created for each token, a matrix of vocab-size x embedding-dimension-size. The model trains to predict if two embedding vectors are related, updates the embeddings, and repeats.

# Recommending songs by embeddings

Let's use the word2vec algorithm to embed songs using human-made music playlists. Think of each song as a token, and the playlist as a sentence. We'll use a dataset of 100s of radio station playlists. Start by loading the dataset.

In [1]:
import pandas as pd
from urllib import request

# Get the playlist dataset file
data = request.urlopen('https://storage.googleapis.com/maps-premium/dataset/yes_complete/train.txt')

# Parse the playlist dataset file. Skip the first two lines as
# they only contain metadata
lines = data.read().decode("utf-8").split('\n')[2:]

# Remove playlists with only one song
playlists = [s.rstrip().split() for s in lines if len(s.split()) > 1]

# Load song metadata
songs_file = request.urlopen('https://storage.googleapis.com/maps-premium/dataset/yes_complete/song_hash.txt')
songs_file = songs_file.read().decode("utf-8").split('\n')
songs = [s.rstrip().split('\t') for s in songs_file]
songs_df = pd.DataFrame(data=songs, columns = ['id', 'title', 'artist'])
songs_df = songs_df.set_index('id')

Each element of the playlist is a list of song IDs.

In [ ]:
print( 'Playlist #1:\n ', playlists[0], '\n')
print( 'Playlist #2:\n ', playlists[1])

Playlist #1:
  ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '2', '42', '43', '44', '45', '46', '47', '48', '20', '49', '8', '50', '51', '52', '53', '54', '55', '56', '57', '25', '58', '59', '60', '61', '62', '3', '63', '64', '65', '66', '46', '47', '67', '2', '48', '68', '69', '70', '57', '50', '71', '72', '53', '73', '25', '74', '59', '20', '46', '75', '76', '77', '59', '20', '43'] 

Playlist #2:
  ['78', '79', '80', '3', '62', '81', '14', '82', '48', '83', '84', '17', '85', '86', '87', '88', '74', '89', '90', '91', '4', '73', '62', '92', '17', '53', '59', '93', '94', '51', '50', '27', '95', '48', '96', '97', '98', '99', '100', '57', '101', '102', '25', '103', '3', '104', '105', '106', '107', '47', '108', '109', '110', '111', '112', '113', '25', '63', '62', '114', '115', '84', '116', '117',

Train the model.

In [2]:
from gensim.models import Word2Vec

# Train our Word2Vec model
model = Word2Vec(
    playlists, vector_size=32, window=20, negative=50, min_count=1, workers=4
)

Use the embeddings to find similar songs. We'll work with "Fade to Black", song id 2172.

In [3]:
song_id = 2172

print(songs_df.iloc[song_id])

title     Fade To Black
artist        Metallica
Name: 2172 , dtype: object


Here are the most similar songs.

In [ ]:
# Ask the model for songs similar to song #2172
model.wv.most_similar(positive=str(song_id))

[('2849', 0.9979680776596069),
 ('2640', 0.9964019060134888),
 ('3167', 0.9963980317115784),
 ('5549', 0.9959008693695068),
 ('2715', 0.9958351850509644),
 ('3117', 0.9954560995101929),
 ('2987', 0.9953479766845703),
 ('2881', 0.9951083660125732),
 ('2886', 0.9950577616691589),
 ('3094', 0.994985044002533)]

In [ ]:
print(songs_df.iloc[2172])

title     Fade To Black
artist        Metallica
Name: 2172 , dtype: object


Let's get their names.

In [5]:
import numpy as np

def print_recommendations(song_id):
    similar_songs = np.array(
        model.wv.most_similar(positive=str(song_id),topn=5)
    )[:,0]
    return  songs_df.iloc[similar_songs]

# Extract recommendations
print_recommendations(2172)

title                               artist
id                                                                 
11517   Mary Had A Little Lamb  Stevie Ray Vaughan & Double Trouble
5634            Mr. Brownstone                        Guns N' Roses
2014           Youth Gone Wild                             Skid Row
5549             November Rain                        Guns N' Roses
11473           Little Guitars                            Van Halen

In [6]:
print_recommendations(842)

title                        artist
id                                                                    
12205                   Give It Up To Me                     Sean Paul
331        Get Low (w\/ Ying Yang Twins)  Lil' Jon & The Eastside Boyz
1560                          In Da Club                       50 Cent
5668             How We Do (w\/ 50 Cent)                      The Game
330     Hate It Or Love It (w\/ 50 Cent)                      The Game